keanu-project/src/main/java/io/improbable/keanu/vertices/dbltensor/Nd4jDoubleTensor.java

```diff
@@ -400,6 +431,16 @@ public DoubleTensor applyWhere(DoubleTensor withMask, double value) {
         return this;
     }
 
+    @Override
+    public DoubleTensor applyInPlace(Function<Double, Double> function) {
```

Comment:
I did a bit of digging into how Nd4j thinks users should do this and found that you can access the data buffer by using the .data() method at which point you effectively have a linear view of the data that you can iterate through.
```
    @Override
    public DoubleTensor applyInPlace(Function<Double, Double> function) {
        DataBuffer data = tensor.data();
        for (int i = 0; i < data.length(); i++) {
            data.put(i, function.apply(data.getDouble(i)));
        }
        return this;
    }
```

Reply:
Also, should add a few tests around this in the Nd4jDoubleTensorTest:

```
    @Test
    public void canApplyUnaryFunctionToScalar() {
        DoubleTensor result = scalarA.apply(a -> a * 2);
        assertEquals(4, result.scalar(), 0.0);
    }

    @Test
    public void canApplyUnaryFunctionToRank3() {
        DoubleTensor rank3Tensor = Nd4jDoubleTensor.create(new double[]{1, 2, 3, 4, 5, 6, 7, 8}, new int[]{2, 2, 2});
        DoubleTensor result = rank3Tensor.apply(a -> a * 2);
        assertArrayEquals(new double[]{2, 4, 6, 8, 10, 12, 14, 16}, result.getLinearView(), 0.0);
    }
```

Reply:
Great, I've made the changes to `applyInPlace` and `apply` plus added in a few tests to Nd4jDoubleTensorTest